### PR TITLE
net/emcute: check value of length field before use

### PR DESCRIPTION
**Pull request by haukepetersen**
The value of the length field was not checked/verified, leading to potential security/stability vulnerabilities. I might not have fixed all problems, but this PR is at least fixing some obvious ones.

-> Thanks to Eric for pointing this out!

### CODE REVIEW

File: sys/net/application_layer/emcute/emcute.c [hunk 1]
```diff
@@ -165,21 +165,24 @@ static void on_ack(uint8_t type, int id_pos, int ret_pos, int res_pos)
     }
 }
 
-static void on_publish(void)
+static void on_publish(uint16_t len, int pos)
 {
+    /* make sure packet length is valid - if not, drop packet silently */
+    if ((int)len < (pos + 6)) {
```
Comment (miri64) [inline sys/net/application_layer/emcute/emcute.c]: Please address Eric's offline comments.

File: sys/net/application_layer/emcute/emcute.c [hunk 2]
```diff
@@ -532,15 +537,25 @@ void emcute_run(uint16_t port, const char *id)
...
             /* handle the packet */
             uint16_t pkt_len;
+            /* catch invalid length field */
+            if ((len == 2) && (rbuf[0] == 0x01)) {
+                continue;
+            }
+            /* parse length field */
             int pos = get_len(rbuf, &pkt_len);
+            /* verify length to prevent overflows */
+            if (((ssize_t)pkt_len > len) || ((ssize_t)pos >= len)) {
```
Comment (miri64) [inline sys/net/application_layer/emcute/emcute.c]: Dito

File: sys/net/application_layer/emcute/emcute.c (hunk 1 above)
Comment (smlng) [inline sys/net/application_layer/emcute/emcute.c]: which were (as I can't read minds over large distances - and for the record)?

Comment (miri64) [inline sys/net/application_layer/emcute/emcute.c]: see security mailing list.

File: sys/net/application_layer/emcute/emcute.c [hunk 3]
```diff
@@ -532,15 +537,25 @@ void emcute_run(uint16_t port, const char *id)
...
+            /* get packet type */
             uint8_t type = rbuf[pos];
 
             switch (type) {
                 case CONNACK:       on_ack(type, 0, 2, 0);  break;
                 case WILLTOPICREQ:  on_ack(type, 0, 0, 0);  break;
                 case WILLMSGREQ:    on_ack(type, 0, 0, 0);  break;
                 case REGACK:        on_ack(type, 4, 6, 2);  break;
-                case PUBLISH:       on_publish();           break;
+                case PUBLISH:       on_publish(len, pos);   break;
```
Comment (smlng) [inline sys/net/application_layer/emcute/emcute.c]: shouldn't this be `on_publish(pkt_len, pos)`?

Comment (haukepetersen) [inline sys/net/application_layer/emcute/emcute.c]: yes, of course. I wonder how that slipped in... Thanks for spotting this!

**Comment by miri64** [comment]
Didn't test yet. Let me do that first :>

**Comment by miri64** [comment]
In the meantime: please squuash

**Comment by miri64** [comment]
Tested. Still works.

**Comment by haukepetersen** [comment]
rebased and squashed.

**Pull request merged**